src/economy/economy.cpp

```diff
@@ -60,8 +60,18 @@ bool nation_is_constructing_factories(sys::state& state, dcon::nation_id n) {
 	// TODO
 	return false;
 }
-bool nation_has_closed_factories(sys::state& state, dcon::nation_id n) {
-	// TODO
+bool nation_has_closed_factories(sys::state& state, dcon::nation_id n) {	// TODO - should be "good" now
+	auto nation_fat = dcon::fatten(state.world, n);
+	for(auto staat_owner : nation_fat.get_state_ownership()) {
```

Comment:
You don't need to iterate over states. Since factories are secretly located in provinces, you can directly iterate over the owned provinces of a nation to find all the factories